### PR TITLE
Windoze/data source refactor

### PR DESCRIPTION
**Pull request by windoze**
Support other data source in DataSource.location

### CODE REVIEW

File: src/test/scala/com/linkedin/feathr/offline/config/location/TestDesLocation.scala [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+          | type: "jdbc"
+          | url: "jdbc:sqlserver://bet-test.database.windows.net:1433;database=bet-test"
+          | user: "bet"
+          | password: "foo"
+          |}""".stripMargin
+      val ds = jackson.readValue(configDoc, classOf[InputLocation])
+      ds match {
+        case Jdbc(url, dbtable, user, password, token, useToken) => {
+          assert(url == "jdbc:sqlserver://bet-test.database.windows.net:1433;database=bet-test")
+          assert(user == "bet")
```
Comment (Yuqing-cat) [inline src/test/scala/com/linkedin/feathr/offline/config/location/TestDesLocation.scala:53]: needs update

File: src/main/scala/com/linkedin/feathr/offline/config/FeathrConfigLoader.scala [hunk 2]
```diff
@@ -712,17 +713,18 @@ private[offline] class DataSourceLoader extends JsonDeserializer[DataSource] {
...
-    val path = if (dataSourceType == "HDFS") {
+    val path: InputLocation = if (dataSourceType == "HDFS") {
       Option(node.get("location")) match {
         case Some(field: ObjectNode) =>
-          Option(field.get("path")) match {
-            case Some(pathField: TextNode) => pathField.textValue()
-            case _ => throw new FeathrConfigException(ErrorLabel.FEATHR_USER_ERROR, "Illegal setting for the path in location")
-          }
+          LocationUtils.getMapper().treeToValue(field, classOf[InputLocation])
+//          Option(field.get("path")) match {
```
Comment (jaymo001) [inline src/main/scala/com/linkedin/feathr/offline/config/FeathrConfigLoader.scala:720]: Delete these.

File: src/main/scala/com/linkedin/feathr/offline/config/location/InputLocation.scala [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, property = "type", defaultImpl = classOf[SimplePath])
+@JsonSubTypes(
+  Array(
+    new JsonSubTypes.Type(value = classOf[SimplePath], name = "path"),
+    new JsonSubTypes.Type(value = classOf[Jdbc], name = "jdbc"),
+  ))
+trait InputLocation {
+  def getPath: String
+
+  def getPathList: Array[String]
```
Comment (jaymo001) [inline src/main/scala/com/linkedin/feathr/offline/config/location/InputLocation.scala:25]: nit: List is preferable.

File: src/main/scala/com/linkedin/feathr/offline/config/location/Jdbc.scala [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+  override def getPathList: Array[String] = Array(url)
+}
+
+object Jdbc {
+  def apply(url: String, dbtable: String, user: String, password: String): Jdbc = Jdbc(url, dbtable, user = user, password = password, useToken = false)
+
+  def apply(url: String, dbtable: String, token: String): Jdbc = Jdbc(url, dbtable, token = token, useToken = true)
+
+  def apply(url: String, dbtable: String): Jdbc = Jdbc(url, dbtable, useToken = false)
```
Comment (jaymo001) [inline src/main/scala/com/linkedin/feathr/offline/config/location/Jdbc.scala:43]: Add doc. How would the authentication work in this case?

File: src/main/scala/com/linkedin/feathr/offline/config/location/PathList.scala [hunk 5]
```diff
@@ -0,0 +1,16 @@
+package com.linkedin.feathr.offline.config.location
+
+import org.apache.spark.sql.{DataFrame, SparkSession}
+
+case class PathList(paths: Array[String]) extends InputLocation {
```
Comment (jaymo001) [inline src/main/scala/com/linkedin/feathr/offline/config/location/PathList.scala:5]: nit: Array -> List

File: src/main/scala/com/linkedin/feathr/offline/config/location/PathList.scala [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+  override def getPath: String = paths.mkString(";")
+
+  override def getPathList: Array[String] = paths
+
+  override def loadDf(ss: SparkSession): DataFrame = ???
```
Comment (jaymo001) [inline src/main/scala/com/linkedin/feathr/offline/config/location/PathList.scala:10]: I think we should load dataframes and union them?

File: src/main/scala/com/linkedin/feathr/offline/config/location/PathList.scala [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+}
+
+object PathList {
+  def apply(path: String): PathList = PathList(path.split(";"))
```
Comment (jaymo001) [inline src/main/scala/com/linkedin/feathr/offline/config/location/PathList.scala:14]: Add doc

File: src/main/scala/com/linkedin/feathr/offline/config/location/PathList.scala [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+  def unapply(o: PathList): Option[Array[String]] = Some(o.paths)
```
Comment (jaymo001) [inline src/main/scala/com/linkedin/feathr/offline/config/location/PathList.scala:15]: nit: Better naming for 'o'

File: src/main/scala/com/linkedin/feathr/offline/config/location/SimplePath.scala [hunk 9]
```diff
@@ -0,0 +1,18 @@
+package com.linkedin.feathr.offline.config.location
+
+import com.fasterxml.jackson.module.caseclass.annotation.CaseClassDeserialize
+import com.linkedin.feathr.offline.source.dataloader.SparkDataLoader
+import org.apache.spark.sql.{DataFrame, SparkSession}
+import org.codehaus.jackson.annotate.JsonProperty
+
+@CaseClassDeserialize()
+case class SimplePath(@JsonProperty("path") path: String = "") extends InputLocation {
```
Comment (jaymo001) [inline src/main/scala/com/linkedin/feathr/offline/config/location/SimplePath.scala:9]: Default value "" is invalid. Do we need it?

File: src/main/scala/com/linkedin/feathr/offline/generation/SparkIOUtils.scala [hunk 10]
```diff
@@ -14,8 +16,17 @@ object SparkIOUtils {
     FileFormat.loadHdfsDataFrame(format, existingHdfsPaths)
   }
 
-  def createDataFrame(path: String, dataIOParams: Map[String, String] = Map()): DataFrame = {
-    createUnionDataFrame(Seq(path), dataIOParams)
+//  def createDataFrame(path: String, dataIOParams: Map[String, String]): DataFrame = {
```
Comment (jaymo001) [inline src/main/scala/com/linkedin/feathr/offline/generation/SparkIOUtils.scala:19]: Delete it

File: src/main/scala/com/linkedin/feathr/offline/generation/SparkIOUtils.scala [hunk 11]
```diff
@@ -14,8 +16,17 @@ object SparkIOUtils {
...
+//    createUnionDataFrame(Seq(path), dataIOParams)
+//  }
+//
+  def createDataFrame(location: InputLocation, dataIOParams: Map[String, String] = Map()): DataFrame = {
+    location match {
+      case SimplePath(path) => createUnionDataFrame(Seq(path), dataIOParams)
+      // TODO:
```
Comment (jaymo001) [inline src/main/scala/com/linkedin/feathr/offline/generation/SparkIOUtils.scala:26]: We should handle default case properly.

File: src/main/scala/com/linkedin/feathr/offline/generation/SparkIOUtils.scala [hunk 12]
```diff
@@ -14,8 +16,17 @@ object SparkIOUtils {
...
+      case Jdbc(url, dbtable, user, password, token, useToken) => JdbcUtils.loadDataFrame(SparkSession.builder.getOrCreate, url)
+    }
+//    createUnionDataFrame(Seq(path), dataIOParams)
```
Comment (jaymo001) [inline src/main/scala/com/linkedin/feathr/offline/generation/SparkIOUtils.scala:29]: Delete

File: src/main/scala/com/linkedin/feathr/offline/util/SourceUtils.scala [hunk 13]
```diff
@@ -679,7 +680,8 @@ private[offline] object SourceUtils {
         if (ss.sparkContext.isLocal){
           getLocalDF(ss, inputData.inputPath)
         } else {
-          loadAsDataFrame(ss, inputData.inputPath)
+          // TODO: InputData needs to be changed to use InputLocation instead of path string
```
Comment (jaymo001) [inline src/main/scala/com/linkedin/feathr/offline/util/SourceUtils.scala:683]: Delete it

**Review by jaymo001 — APPROVED** [review]

**Review by xiaoyongzhu — APPROVED** [review]

**Pull request merged**